File: PluginChangesReport.php
<?php
# Script to generate report of unreleased plugin changes in Jenkins' plugin repositories.
# @author Alan Harder (mindless@dev.java.net)
$time = time();
#
# $knownRevs of $id-$ver-$cnt => text
# To display a message in the report regarding the state of a plugin.
# Usually to show a message like "~CURRENT -- pom,test" if a plugin's
# changes since the last release don't actually change anything in a
# release (such as pom changes or unit tests).
# For plugin $id with $cnt revs since version $ver, show given message.
# Or $id-unreleased to show given message for an unreleased plugin.
$knownRevs = readFromStdin();
#
# $repoMap of $id => repo
# To handle cases where artifactId from POM doesn't map directly
# to github repo name or directory name in svn. Set to 'skip' to skip item.
# Default for repo is $id-plugin for github and $id for svn.
$repoMap = readFromStdin();
#
# $reallyGithub of $id=>1
# For plugins that migrate to github but have not yet run a release there.
$reallyGithub = readFromStdin();
#
$prefix = isset($argv[1]) ? $argv[1] : '';
$issueUrl = 'http://issues.jenkins-ci.org/browse';
$svn = 'svn --non-interactive';
$svnBase = 'https://svn.jenkins-ci.org';
# Where to find all svn tags (a couple plugins use a subdir of /tags)
$svnTagDirs = array("$svnBase/tags",
                    "$svnBase/tags/global-build-stats",
                    "$svnBase/tags/scm-sync-configuration");
$fisheyeBase = 'http://fisheye.jenkins-ci.org';
$fisheyeUrl = array("$fisheyeBase/search/Jenkins/trunk/hudson/plugins/",
                    '?ql=select%20revisions%20from%20dir%20/trunk/hudson/plugins/',
                    '%20where%20date%20%3E=%20',
                    '%20group%20by%20changeset%20return%20csid,%20comment,%20author,%20path');
$monthMap = array('Jan'=>1, 'Feb'=>2, 'Mar'=>3, 'Apr'=>4, 'May'=>5, 'Jun'=>6,
                  'Jul'=>7, 'Aug'=>8, 'Sep'=>9, 'Oct'=>10, 'Nov'=>11, 'Dec'=>12);
date_default_timezone_set('America/Los_Angeles');

$seenGithubRepos = $seenJavanetDirs = $svnTagMap = array();

# 1. Load update-center.json
$updateCenter = json_decode(
        trim(file_get_contents('http://updates.jenkins-ci.org/update-center.json'),
             "updateCnr.os(); \t\n\r"));
if (!$updateCenter) mydie('** No data from update-center.json');

# 2. Load all tags from svn
$xml = xml_parser_create();
$svnTagDirs = implode(' ', $svnTagDirs);
xml_parse_into_struct($xml,
    `$svn ls --xml $svnTagDirs`, $xmlData, $xmlIndex);
xml_parser_free($xml);
if (!$xmlData) mydie('** Failed to get tags from svn');
foreach ($xmlIndex['NAME'] as $i) {
  $tag = $xmlData[$i]['value'];
  $rev = $xmlData[$i+2]['attributes']['REVISION'];
  # Split on last - to get plugin id and version#
  $p = ($i = strrpos($tag, '-')) === FALSE ? $tag : substr($tag, 0, $i);
  $v = $i === FALSE ? '' : substr($tag, $i + 1);
  # Build map of id->(version#,rev#), keeping latest rev# for each plugin
  if (!isset($svnTagMap[$p]) or $rev > $svnTagMap[$p][1]) $svnTagMap[$p] = array($v, $rev);
}

# 3. Process all released plugins
$data = array();
foreach ($updateCenter->plugins as $id => $p) {
  if ($prefix and !preg_match("/^$prefix/i", $id)) continue;
  $isGithub = strcasecmp(substr(isset($p->scm) ? $p->scm : '', -10), 'github.com') === 0
           || isset($reallyGithub[$id]);
  $repoName = isset($repoMap[$id]) ? $repoMap[$id] : ($isGithub ? "$id-plugin" : $id);
  if ($repoName==='skip') continue;
  # Check either github or svn for revisions since the latest release.
  if ($isGithub) {
    $seenGithubRepos[$repoName] = 1;
    $seenJavanetDirs[$id] = 1; # In case moved to github from svn
    list($ver, $revs, $url) = github($id, $repoName);
  } else {
    $seenJavanetDirs[$repoName] = 1;
    $seenGithubRepos[$repoName . '-plugin'] = 1; # Skip github mirror too
    list($ver, $revs, $url) = jenkinsSvn($id, $repoName, $p);
  }
  # Examine/count the revs-since-release and print resulting data.
  $data[] = processRevs($revs, $id, $p, $ver, $url);
}

# Skip github mirrors of svn dirs too
foreach ($repoMap as $key => $value) if ($value=='skip') $repoMap["$key-plugin"] = 'skip';

# 4. Get list of all directories under hudson/plugins in svn
#    and report any unreleased plugins
exec("$svn ls $svnBase/trunk/hudson/plugins", $plugins);
if (!$plugins) mydie('** Failed to get plugin list from svn');
foreach ($plugins as $p) {
  if ($prefix and !preg_match("/^$prefix/i", $p)) continue;
  if (substr($p, -1) == '/') {
    $p = substr($p, 0, -1);
    if (isset($repoMap[$p]) and $repoMap[$p] === 'skip') continue;
    if (!isset($seenJavanetDirs[$p])) {
      # TODO: get #revs and date range..
      $comment = knownRevs($p . '-unreleased');
      if (!$comment) $comment = 'unreleased';
      $data[] = "| [$p|$fisheyeBase/browse/Jenkins/trunk/hudson/plugins/$p] | | | $comment\n";
      $seenGithubRepos["$p-plugin"] = 1; # Don't also report github mirror as unreleased
    }
  }
}
# 5. Get list of all git repositories under github.com/jenkinsci
#    and report any unreleased plugins
for ($p = 1; TRUE; $p++) {
  $githubRepos = json_decode(
        file_get_contents('http://github.com/api/v2/json/repos/show/jenkinsci?page=' . $p));
  if (!$githubRepos) mydie('** Failed to get repo list from github');
  if (count($githubRepos->repositories) == 0) break;
  foreach ($githubRepos->repositories as $repo) {
    if ($prefix and !preg_match("/^$prefix/i", $repo->name)) continue;
    if (isset($repoMap[$repo->name]) and $repoMap[$repo->name] === 'skip') continue;
    if (!isset($seenGithubRepos[$repo->name])) {
      # TODO: get #revs and date range..
      $comment = knownRevs($repo->name . '-unreleased');
      if (!$comment) $comment = 'unreleased';
      $data[] = "| [$repo->name|http://github.com/jenkinsci/$repo->name] | | | $comment\n";
    }
  }
}

# 6. Group and print results
usort($data, function($a,$b){return strcasecmp(ltrim($a,'| ['),ltrim($b,'| ['));});
print "This is a report of unreleased changes in Jenkins' plugin repositories.\n"
    . "It is updated once per week.\n\nh3. Plugin Changes\n";
foreach ($data as $line) {
  if (strpos($line, 'CURRENT') === FALSE and strpos($line, 'unreleased') === FALSE) print $line;
}
print "\nh3. Unreleased Plugins\n";
foreach ($data as $line) {
  if (strpos($line, 'unreleased') !== FALSE) print $line;
}
print "\nh3. Current Plugins\n";
foreach ($data as $line) {
  if (strpos($line, 'CURRENT') !== FALSE) print $line;
}

# 7. Report any unused $knownRevs entries
foreach ($knownRevs as $key => $value) {
  if ($prefix and !preg_match("/^$prefix/i", $key)) continue;
  $data[] = "| Unused data in KnownRevs: | | | | $key | $value\n";
}
$time = time() - $time;
print "\nGenerated at: " . `date` . ' in ' . floor($time/60) . ' min ' . ($time%60) . " sec.\n";

### Helper methods

function knownRevs($key) {
  global $knownRevs;
  if (isset($knownRevs[$key])) {
    $result = $knownRevs[$key];
    unset($knownRevs[$key]);
  } else $result = FALSE;
  return $result;
}

function github($pluginId, $repoName) {
  # Get all tags in this repo, sort by version# and get highest
  list ($ver, $hash) = maxTag($pluginId, json_decode(
    file_get_contents("http://github.com/api/v2/json/repos/show/jenkinsci/$repoName/tags")));
  $revs = array();
  # URL to compare last release tag and master branch
  $url = "https://github.com/jenkinsci/$repoName/compare/$hash...master";
  # Fetch ".patch" version of this URL and split into revisions
  foreach (explode("\nFrom ", file_get_contents("$url.patch")) as $rev) {
    if (!preg_match(
          '|^Date: \w{3}, (\d+ \w+ \d+).*?\nSubject: \[PATCH[ \d/]*\]\s*(.*?)$|m', $rev, $match)) {
      mydie("** Failed to parse github revision for $pluginId: $rev");
    }
    $revs[] = array(dateFormat($match[1]), $match[2]);
  }
  return array($ver, $revs, $url);
}

function maxTag($pluginId, $json) {
  $pidLen = strlen($pluginId);
  $vers = array();
  foreach ($json->tags as $id => $hash) {
    if (preg_match("/^(?:$pluginId" . '-?)?([0-9._]+)$/', $id, $match))
      $vers[$match[1]] = $hash;
    else fwrite(STDERR, "** Skipped github tag: $pluginId - $id\n");
  }
  uksort($vers, 'version_compare');
  return each(array_reverse($vers));
}

function dateFormat($date) {
  global $monthMap;
  return preg_match('/^(\d+) (\w+) (\d+)$/', $date, $match)
    ? sprintf('%d-%02d-%02d', $match[3], $monthMap[$match[2]], $match[1]) : $date;
}

function jenkinsSvn($pluginId, $repoName, $pluginJson) {
  global $fisheyeUrl, $svnTagMap, $svn, $svnBase, $knownRevs;
  # URL for changes since last release
  $url = "$fisheyeUrl[0]$repoName$fisheyeUrl[1]$repoName$fisheyeUrl[2]"
    . $pluginJson->releaseTimestamp . $fisheyeUrl[3];
  $revs = array();
  # Get rev# when tag for last release was created
  list ($ver, $tagRev) =
    isset($svnTagMap[$pluginId]) ? $svnTagMap[$pluginId]
      : (isset($svnTagMap[$repoName]) ? $svnTagMap[$repoName] : $svnTagMap["$pluginId-plugin"]);
  if (!$tagRev) {
    $ver = '?';
    $key = "$pluginId-$ver-0";
    if (!isset($knownRevs[$key])) $knownRevs[$key] = '?'; # Don't show "CURRENT"
    fwrite(STDERR, "** Unable to find latest svn tag for $pluginId\n");
  } else {
    # Get revisions in trunk since last release
    $xml = xml_parser_create();
    xml_parse_into_struct($xml,
      `$svn log -r $tagRev:HEAD --xml $svnBase/trunk/hudson/plugins/$repoName`,
      $xmlData, $xmlIndex);
    xml_parser_free($xml);
    if (isset($xmlIndex['MSG'])) {
      foreach ($xmlIndex['MSG'] as $i) {
        $revs[] = array(substr($xmlData[$i-2/*DATE*/]['value'], 0, 10),
                        isset($xmlData[$i/*MSG*/]['value']) ? $xmlData[$i]['value'] : '');
      }
    } else {
      $key = "$pluginId-$ver-0";
      if (!isset($knownRevs[$key])) $knownRevs[$key] = '?'; # Don't show "CURRENT"
      fwrite(STDERR, "** Unable to find revisions for hudson/plugins/$repoName from r$tagRev\n");
    }
  }
  return array($ver, $revs, $url);
}

function processRevs($revs, $pluginId, $pluginJson, $ver, $url) {
  global $knownRevs, $issueUrl;
  $cnt = $l10n = 0;
  $firstDate = $result = FALSE;
  $fixed = array();
  foreach ($revs as $rev) {
    list ($date, $comment) = $rev;
    # Skip these commits from [maven-release]
    if (strpos($comment, 'prepare for next development iteration') !== FALSE
        or strncmp($comment, 'bumping up POM version', 22) === 0) {
      continue;
    }
    $cnt++;
    if (!$firstDate) $firstDate = $date;
    if (preg_match(
          '/^(integrated )?community[- ]contributed (localization|translation)/i', $comment)) {
      $l10n++;
    }
    for ($i = 0; preg_match('/FIXED ([A-Z]+-(\d+))/i',
                            $comment, $match, PREG_OFFSET_CAPTURE, $i);) {
      $fixed[] = '[' . $match[2][0] . "|$issueUrl/" . $match[1][0] . ']';
      $i = $match[2][1] + strlen($match[2][0]);
    }
  }

  $key = "$pluginId-$ver-$cnt";
  if (isset($knownRevs[$key])) { $result = $knownRevs[$key]; unset($knownRevs[$key]); }
  elseif (!$cnt) $result = 'CURRENT';
  else {
    if ($fixed) $result = 'Fixed: ' . implode(' ', $fixed);
    elseif ($l10n == $cnt) $result = '~CURRENT --';
    if ($l10n) $result .= ' l10n';
  }
  if ($ver and $ver !== $pluginJson->version)
    $result .= ' (_Version mismatch: json has ' . $pluginJson->version . '_)';

  $since = $cnt ? "[$cnt rev" . ($cnt > 1 ? 's' :'') . "|$url] | since" : '|';
  $p = !empty($pluginJson->wiki) ? "[$pluginId|$pluginJson->wiki]" : $pluginId;
  $today = today();
  $d = $firstDate ? ($firstDate==$date ? colorize($date, $today)
     : colorize($firstDate, $today) . ' to ' . colorize($date, $today)) : '';
  return "| $p | $since $ver | $d | $result\n";
}


function today() {
  $x = localtime();
  return $x[5] * 12 + $x[4] + $x[3]/31;
}

function ageMonths($date, $today) {
  $x = explode('-', $date);
  return $today - (($x[0] - 1900) * 12 + ($x[1] - 1) + $x[2]/31);
}

function colorize($date, $today) {
  $a = round(ageMonths($date,$today)/2);
  if ($a <= 0) return $date;
  if ($a > 5) $a = 5;
  $colors = array(0, '3', '6', '9', 'c', 'f');
  return '{color:#' . $colors[$a] . (9 - $a) . '6}' . $date . '{color}';
}

function readFromStdin() {
  $result = array();
  while (($line = fgets(STDIN)) !== FALSE and strncmp($line, '----', 4)) {
    if (($i = strpos($line, '#')) !== FALSE) $line = substr($line, 0, $i);
    $line = trim($line);
    if (!$line) continue;
    if (($i = strpos($line, '|')) > 0) {
      $result[rtrim(substr($line, 0, $i))] = ltrim(substr($line, $i + 1));
    } else {
      $result[$line] = 1;
    }
  }
  return $result;
}

function mydie($msg) {
  fwrite(STDERR, "$msg\n");
  exit(1);
}
?>

File: PluginChangesReport.py
#!/usr/bin/python
# Script to generate report of unreleased plugin changes in Jenkins' plugin repositories.
# @author Alan Harder
#
from distutils import version
import json
import os
import re
from StringIO import StringIO
import sys
from time import localtime, sleep, time
from datetime import datetime
from urllib2 import urlopen
from xml.etree import ElementTree

start_time = time()

# $knownRevs of $id-$ver-$cnt => text
# To display a message in the report regarding the state of a plugin.
# Usually to show a message like "~CURRENT -- pom,test" if a plugin's
# changes since the last release don't actually change anything in a
# release (such as pom changes or unit tests).
# For plugin $id with $cnt revs since version $ver, show given message.
# Or $id-unreleased to show given message for an unreleased plugin.
knownRevs = {}

# $repoMap of $id => repo
# To handle cases where artifactId from POM doesn't map directly
# to github repo name or directory name in svn. Set to 'skip' to skip item.
# Default for repo is $id-plugin for github and $id for svn.
repoMap = {}

# $tagMap of $id => tagBase | [pluginSubDir or VER_OK] [ | suffix ]
# Assists in finding latest version of a plugin when the tags are not
# simple "pluginId-version" due to non-standard tags or use of a parent pom.
# Maps pluginId to the base name used for tags and how to find the release
# version; the latter can be "VER_OK" which means the version in the tag
# matches the release version; if the parent pom has different version
# numbers than the plugin release, specify the subdirectory in the source
# where the plugin resides and the version will be looked up in the pom.
# Optionally add |suffix to specify a regex to allow after the version#.
tagMap = {}

# $reallyGithub of $id=>1
# For plugins that migrate to github but have not yet run a release there.
reallyGithub = {}

issueUrl = 'http://issues.jenkins-ci.org/browse'
svn = 'svn --non-interactive'
svnBase = 'https://svn.jenkins-ci.org'
# Where to find all svn tags (a couple plugins use a subdir of /tags)
svnTagDirs = [ svnBase + '/tags',
               svnBase + '/tags/global-build-stats',
               svnBase + '/tags/scm-sync-configuration' ]
fisheyeBase = 'http://fisheye.jenkins-ci.org'
fisheyeUrl = [ fisheyeBase + '/search/Jenkins/trunk/hudson/plugins/',
               '?ql=select%20revisions%20from%20dir%20/trunk/hudson/plugins/',
               '%20where%20date%20%3E=%20',
               '%20group%20by%20changeset%20return%20csid,%20comment,%20author,%20path' ]
monthMap = { 'Jan': 1, 'Feb': 2, 'Mar': 3, 'Apr': 4, 'May': 5, 'Jun': 6,
             'Jul': 7, 'Aug': 8, 'Sep': 9, 'Oct': 10, 'Nov': 11, 'Dec': 12 }
svnTagMap = {}

def main():
  prefix = sys.argv[1] if len(sys.argv) > 1 else ''

  # 1. Load update-center.json
  updateCenter = json.load(StringIO(
        urlopen('http://updates.jenkins-ci.org/update-center.json')
            .read().strip('updateCnr.os(); \t\n\r')))

  # 2. Load all tags from svn
  xml = ElementTree.XML(
        os.popen('%s ls --xml %s' % (svn, ' '.join(svnTagDirs))).read())
  for entry in xml.findall('list/entry'):
    tag = entry.find('name').text
    rev = int(entry.find('commit').get('revision'))
    if '-' in tag:
      # Split on last - to get plugin id and version#
      (p, v) = tag.rsplit('-', 1)
      # Build map of id->(version#,rev#), keeping latest rev# for each plugin
      if p not in svnTagMap or rev > svnTagMap[p][1]: svnTagMap[p] = (v, rev)

  # 3. Process all released plugins
  data = []
  seenGithubRepos = {}
  seenJavanetDirs = {}
  for (id, p) in updateCenter['plugins'].items():
    if not id.startswith(prefix): continue
    if 'scm' not in p: print >> sys.stderr, '** Missing scm info for', id
    isGithub = ('scm' in p and p['scm'][-10:].lower() == 'github.com'
                or id in reallyGithub)
    # By default $id matches dir name in svn, and "$id-plugin" is repo name in github:
    repoName = (repoMap[id] if id in repoMap
                else (id + '-plugin' if isGithub and id[-7:] != '-plugin' else id))
    if repoName == 'skip': continue
    # Check either github or svn for revisions since the latest release.
    if isGithub:
      seenGithubRepos[repoName] = 1
      seenJavanetDirs[id] = 1  # In case moved to github from svn
      (ver, revs, url) = github(id, repoName)
    else:
      seenJavanetDirs[repoName] = 1
      seenGithubRepos[repoName + '-plugin'] = 1  # Skip github mirror too
      (ver, revs, url) = jenkinsSvn(id, repoName, p)
    # Examine/count the revs-since-release and add to data.
    data.append(processRevs(revs, id, p, ver, url))

  # Skip github mirrors of svn dirs too
  for (key, value) in repoMap.items():
    if value == 'skip': repoMap[key + '-plugin'] = 'skip'

  # 4. Get list of all directories under hudson/plugins in svn
  #    and report any unreleased plugins
  for p in [ x.rstrip() for x in
             os.popen('%s ls %s/trunk/hudson/plugins' % (svn, svnBase)) ]:
    if not p.startswith(prefix): continue
    if p[-1] == '/':
      p = p[:-1]
      if p in repoMap and repoMap[p] == 'skip': continue
      if p not in seenJavanetDirs:
        # TODO: get #revs and date range..
        comment = getKnownRevs(p + '-unreleased')
        if not comment: comment = 'unreleased'
        data.append('| [%s|%s/browse/Jenkins/trunk/hudson/plugins/%s] | | | %s'
                    % (p, fisheyeBase, p, comment))
        seenGithubRepos[p + '-plugin'] = 1  # Don't also report github mirror as unreleased

  # 5. Get list of all git repositories under github.com/jenkinsci
  #    and report any unreleased plugins
  page = 1
  while True:
    githubRepos = json.load(
        urlopen('https://api.github.com/orgs/jenkinsci/repos?page=%s&per_page=100' % page))
    if len(githubRepos) == 0: break
    for repoName in [ repo['name'] for repo in githubRepos ]:
      if not repoName.startswith(prefix): continue
      if repoName in repoMap and repoMap[repoName] == 'skip': continue
      if repoName not in seenGithubRepos:
        # TODO: get #revs and date range..
        comment = getKnownRevs(repoName + '-unreleased')
        if not comment: comment = 'unreleased'
        data.append('| [%s|http://github.com/jenkinsci/%s] | | | %s'
                    % (repoName, repoName, comment))
    page += 1

  # 6. Group and print results
  data = sorted(data, key=lambda s: s.lstrip('| [').lower())
  print('This is a report of unreleased changes in Jenkins\' plugin repositories.\n'
        'It is updated once per week.\n\nh3. Plugin Changes')
  for line in data:
    if not 'CURRENT' in line and not 'unreleased' in line: print line
  print '\nh3. Unreleased Plugins'
  for line in data:
    if 'unreleased' in line: print line
  print '\nh3. Current Plugins'
  for line in data:
    if 'CURRENT' in line: print line

  # 7. Report any unused $knownRevs entries
  for (key, value) in knownRevs.items():
    if not key.startswith(prefix): continue
    print '| Unused data | in | KnownRevs: | %s | %s' % (key, value);

  duration = int(time() - start_time)
  print('\nGenerated at: %s in %d min %d sec\n'
        % (os.popen('date').read(), duration/60, duration%60))

### Helper methods

def readFromStdin():
  result = {}
  for line in sys.stdin:
    if line.startswith('----'): break
    if '#' in line: line = line[:line.index('#')]
    line = line.strip()
    if not line: continue
    if '|' in line:
      result[line[:line.index('|')].rstrip()] = line[line.index('|')+1:].lstrip()
    else:
      result[line] = 1
  return result

def getUrl(url):
  sleep(3)  # Don't hit github too fast
  try:
    return urlopen(url)
  except IOError as e:
    # Retry once
    sleep(3)
    print >> sys.stderr, '**', e, '\n** Retry', url
    try: return urlopen(url)
    except IOError as e:
      print >> sys.stderr, '**', e
      return False

def getJson(url):
  s = getUrl(url)
  return json.load(s) if s else False

def getKnownRevs(key):
  if key in knownRevs:
    result = knownRevs[key]
    del knownRevs[key]
  else: result = False
  return result;

def github(pluginId, repoName):
  # Prepend github account "jenkinsci" if another account not specified
  if '/' not in repoName: repoName = 'jenkinsci/' + repoName
  # Get all tags in this repo, sort by version# and get highest
  (ver, tag) = maxTag(pluginId, repoName, getJson(
        'https://api.github.com/repos/%s/tags' % repoName))
  revs = getRevs(repoName, tag)
  # URL to compare last release tag and master branch
  url = 'https://github.com/%s/compare/%s...master' % (repoName, tag['name'])
  return (ver, revs, url)

def getRevs(repoName, tag):
  commits = getJson('https://api.github.com/repos/%s/commits' % repoName)
  revs = []
  if not commits: return revs
  for commit in commits:
    revs.append( (dateFormat2(commit['commit']['author']['date']), commit['commit']['message']) )
    if commit['sha'] == tag['commit']['sha']:
      break
  return revs;

def maxTag(pluginId, repoName, json):
  if not json: return ('', '')
  vers = {}
  rx = re.compile('^(?:%s-?)?([0-9._]+)$' % pluginId)
  for tagentry in json:
    tag = tagentry['name']
    match = rx.match(tag)
    if match:
      vers[match.group(1)] = tagentry
      continue
    elif pluginId in tagMap:
      entry = tagMap[pluginId].split('|')
      if len(entry) < 3: entry.append('')
      match = re.match('^(?:%s-?)?([0-9._]+%s)$' % (entry[0], entry[2]), tag)
      if match:
        ver = lookupVersion(entry[1], match.group(1), repoName, tag)
        if ver:
          vers[ver] = tagentry
          continue
    print >> sys.stderr, '** Skipped github tag: %s - %s' % (pluginId, tag)
  if not vers: return ('', '')
  key = max(vers.keys(), key=lambda v: version.LooseVersion(v.replace('_', '.')))
  return (key, vers[key])

def lookupVersion(pluginSubDir, tagVersion, repoName, tag):
  if pluginSubDir == 'VER_OK': return tagVersion
  xml = ElementTree.XML(
      urlopen('https://github.com/%s/raw/%s/%s/pom.xml'
              % (repoName, tag, pluginSubDir)).read())
  # Need to account for xmlns in find:
  return xml.find('{%s}version' % xml.tag[1:xml.tag.index('}')]).text

def dateFormat(date):
  match = re.match('^(\d+) (\w+) (\d+)$', date)
  return ('%s-%02d-%02d' % (match.group(3), monthMap[match.group(2)], int(match.group(1)))
          if match else date)

def dateFormat2(date):
  dt=datetime.strptime(date, '%Y-%m-%dT%H:%M:%SZ')
  return dt.strftime('%Y-%m-%d')

def jenkinsSvn(pluginId, repoName, pluginJson):
  # URL for changes since last release
  url = (fisheyeUrl[0] + repoName + fisheyeUrl[1] + repoName + fisheyeUrl[2]
         + pluginJson['releaseTimestamp'] + fisheyeUrl[3])
  revs = []
  # Get rev# when tag for last release was created
  (ver, tagRev) = (svnTagMap[pluginId] if pluginId in svnTagMap else
      (svnTagMap[repoName] if repoName in svnTagMap else svnTagMap[pluginId + '-plugin']))
  if not tagRev:
    ver = '?'
    key = '%s-%s-0' % (pluginId, ver)
    if key not in knownRevs: knownRevs[key] = '?'  # Don't show "CURRENT"
    print >> sys.stderr, '** Unable to find latest svn tag for %s' % pluginId
  else:
    # Get revisions in trunk since last release
    xml = ElementTree.XML(
        os.popen('%s log -r %s:HEAD --xml %s/trunk/hudson/plugins/%s'
                 % (svn, tagRev, svnBase, repoName)).read())
    log = xml.findall('logentry')
    if log:
      for entry in log:
        msg = entry.find('msg').text
        revs.append( (entry.find('date').text[:10],
                      msg if msg is not None else '') )
    else:
      key = '%s-%s-0' % (pluginId, ver)
      if key not in knownRevs: knownRevs[key] = '?'  # Don't show "CURRENT"
      print >> sys.stderr, ('** Unable to find revisions for hudson/plugins/%s from r%s'
                            % (repoName, tagRev))
  return (ver, revs, url)

def processRevs(revs, pluginId, pluginJson, ver, url):
  cnt = 0
  l10n = 0
  l10n_rx = re.compile(
      '^(integrated )?community[- ]contributed (localization|translation)', re.IGNORECASE)
  issue_rx = re.compile('FIX[EDS]* [A-Z]+-(\d+)', re.IGNORECASE)
  firstDate = False
  result = ''
  fixed = []
  for (date, comment) in revs:
    # Skip these commits from [maven-release]
    if ('prepare for next development' in comment
        or comment.startswith('bumping up POM version')):
      continue
    cnt += 1
    if not firstDate: firstDate = date
    if l10n_rx.match(comment): l10n += 1
    for match in issue_rx.findall(comment):
      fixed.append('[%s|%s/JENKINS-%s]' % (match, issueUrl, match))

  key = '%s-%s-%d' % (pluginId, ver, cnt)
  if key in knownRevs:
    result = knownRevs[key]
    del knownRevs[key]
  elif not cnt: result = 'CURRENT'
  else:
    if fixed: result = 'Fixed: ' + ' '.join(fixed)
    elif l10n == cnt: result = '~CURRENT --'
    if l10n: result += ' l10n'
  if ver and ver != pluginJson['version']:
    result += ' (_Version mismatch: json has %s_)' % pluginJson['version']

  since = '[%d rev%s|%s] | since' % (cnt, 's' if cnt > 1 else '', url) if cnt else '|'
  p = '[%s|%s]' % (pluginId, pluginJson['wiki']) if 'wiki' in pluginJson else pluginId
  today = getToday()
  if firstDate:
    d = (colorize(date, today) if firstDate == date else
         colorize(firstDate, today) + ' to ' + colorize(date, today))
  else: d = ''
  return '| %s | %s %s | %s | %s' % (p, since, ver, d, result)

def getToday():
  t = localtime()
  return (t.tm_year - 1900) * 12 + t.tm_mon + t.tm_mday/31.0;

def ageMonths(date, today):
  x = date.split('-')
  return today - ((int(x[0]) - 1900) * 12 + int(x[1]) + int(x[2])/31.0)

def colorize(date, today):
  a = int(round(ageMonths(date, today)/2.0))
  if a <= 0: return date
  if a > 5: a = 5
  colors = [ 0, '3', '6', '9', 'c', 'f' ]
  return '{color:#%s%d6}%s{color}' % (colors[a], 9 - a, date)

knownRevs = readFromStdin()
repoMap = readFromStdin()
tagMap = readFromStdin()
reallyGithub = readFromStdin()

main()
